Peer dependencies are missing in next.js build package.json

<!-- Please do your best to fill out all of the sections below! -->

## Current Behavior
<!-- What is the behavior that currently you experience? -->

When building next.js applications, peer dependencies of dependencies that are installed are missing in the generated package.json file. Two examples:

To use `@apollo/client`, one need to also install `graphql` as it is a peerDependency of `@apollo/client`
```shell
yarn add @apollo/client graphql
```

In the code, `graphql` will never be required, resulting in it not being added to `package.json` in the build.

The same goes for `@emotion/styled`, it requires a peer dependency of `@emotion/react`, which is never required in the code and thus, is missing in the package.json.

Those are only two examples that I have in my app but there are countless others.

## Expected Behavior
<!-- What is the behavior that you expect to happen? -->
<!-- Is this a regression? .i.e Did this used to be the behavior at one point?  -->

I expect the `package.json` to include peerDependencies or to allow me to specify which extra dependencies I want.

## Steps to Reproduce
<!-- Help us help you by making it easy for us to reproduce your issue! -->

<!-- Can you reproduce this on https://github.com/nrwl/nx-examples? -->
<!-- If so, open a PR with your changes and link it below. -->
<!-- If not, please provide a minimal Github repo -->
<!-- At the very least, provide as much detail as possible to help us reproduce the issue -->

```shell
npx create-nx-workspace depexample --preset empty --nx-cloud 0
cd depexample
yarn remove @nrwl/workspace --dev
yarn add @nrwl/next @nrwl/workspace
nx g @nrwl/next:application --name nextapp --style @emotion/styled
yarn add @apollo/client graphql
echo "import '@apollo/client'" >> apps/nextapp/pages/_app.tsx
nx build nextapp
```

To mimic running in a container, just move the build folder to outside the Monorepo and do a yarn install.

### Failure Logs
<!-- Please include any relevant log snippets or files here. -->

And when you run the application:

```
yarn start
```

```
Error: Cannot find module 'graphql'
Require stack:
```

### Environment
<!-- It's important for us to know the context in which you experience this behavior! -->
<!-- Please paste the result of `nx report` below! -->

```
>  NX  Report complete - copy this into the issue template

  Node : 14.4.0
  OS   : darwin x64
  yarn : 1.22.4
  
  nx : Not Found
  @nrwl/angular : Not Found
  @nrwl/cli : 11.1.2
  @nrwl/cypress : 11.1.2
  @nrwl/devkit : 11.1.2
  @nrwl/eslint-plugin-nx : 11.1.2
  @nrwl/express : Not Found
  @nrwl/jest : 11.1.2
  @nrwl/linter : 11.1.2
  @nrwl/nest : Not Found
  @nrwl/next : 11.1.2
  @nrwl/node : Not Found
  @nrwl/react : 11.1.2
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.1.2
  @nrwl/web : 11.1.2
  @nrwl/workspace : 11.1.2
  typescript : 4.0.5
```

--- a/e2e/node/src/node.test.ts
+++ b/e2e/node/src/node.test.ts
@@ -185,6 +185,9 @@ describe('Build Node apps', () => {
         dependencies: {
           '@nestjs/common': '^7.0.0',
           '@nestjs/core': '^7.0.0',
+          '@nestjs/platform-express': '^7.0.0',
+          'reflect-metadata': '^0.1.13',
+          rxjs: '~6.6.3',
         },
         main: 'main.js',
         name: expect.any(String),

--- a/packages/next/src/executors/build/lib/create-package-json.ts
+++ b/packages/next/src/executors/build/lib/create-package-json.ts
@@ -1,76 +1,32 @@
 import { ExecutorContext } from '@nrwl/devkit';
 
-import { writeFileSync } from 'fs';
-import { join } from 'path';
-
 import { createProjectGraph } from '@nrwl/workspace/src/core/project-graph';
-import { readJsonFile } from '@nrwl/workspace/src/utilities/fileutils';
-import { calculateProjectDependencies } from '@nrwl/workspace/src/utilities/buildable-libs-utils';
+import { writeJsonFile } from '@nrwl/workspace/src/utilities/fileutils';
+import { createPackageJson as generatePackageJson } from '@nrwl/workspace/src/utilities/create-package-json';
 
 import { NextBuildBuilderOptions } from '../../../utils/types';
-
-function getProjectDeps(context: ExecutorContext, rootPackageJson: any) {
-  const projGraph = createProjectGraph();
-  const { dependencies: deps } = calculateProjectDependencies(
-    projGraph,
-    context.root,
-    context.projectName,
-    context.targetName,
-    context.configurationName
-  );
-  const depNames = deps
-    .map((d) => d.node)
-    .filter((node) => node.type === 'npm')
-    .map((node) => node.data.packageName)
-    // Need to make sure @nrwl/workspace is installed
-    // It is only a peer dependency of @nrwl/next so does not get installed automatically
-    // See: https://github.com/nrwl/nx/issues/4336
-    .concat('@nrwl/workspace');
-  const dependencies: string[] = depNames
-    .filter((packageName) => packageName in rootPackageJson.dependencies)
-    .reduce((deps, pkgName) => {
-      return { ...deps, [pkgName]: rootPackageJson.dependencies[pkgName] };
-    }, {});
-  const devDependencies = depNames
-    .filter((packageName) => packageName in rootPackageJson.devDependencies)
-    .reduce((deps, pkgName) => {
-      return { ...deps, [pkgName]: rootPackageJson.devDependencies[pkgName] };
-    }, {});
-  return {
-    dependencies,
-    devDependencies,
-  };
-}
 
 export function createPackageJson(
   options: NextBuildBuilderOptions,
   context: ExecutorContext
 ) {
-  const rootPackageJson = readJsonFile(join(context.root, 'package.json'));
-  const { dependencies, devDependencies } = getProjectDeps(
-    context,
-    rootPackageJson
-  );
+  const depGraph = createProjectGraph();
+  const packageJson = generatePackageJson(context.projectName, depGraph, {
+    root: context.root,
+    projectRoot: context.workspace.projects[context.projectName].sourceRoot,
+  });
+  if (!packageJson.scripts) {
+    packageJson.scripts = {};
+  }
+  packageJson.scripts.start = 'next start';
+  if (!packageJson.devDependencies) {
+    packageJson.devDependencies = {};
+  }
+  const nrwlWorkspaceNode = depGraph.nodes['npm:@nrwl/workspace'];
 
-  const outPackageJson = {
-    name: context.projectName,
-    version: '0.0.1',
-    scripts: {
-      start: 'next start',
-    },
-    dependencies: {
-      ...dependencies,
-      // peer deps of next, so we need to add them here
-      react: rootPackageJson.dependencies['react'],
-      'react-dom': rootPackageJson.dependencies['react-dom'],
-      next: rootPackageJson.dependencies['next'],
-    },
-    // needed for the next.config.js file
-    devDependencies,
-  };
-
-  writeFileSync(
-    join(options.outputPath, 'package.json'),
-    JSON.stringify(outPackageJson, null, 2)
-  );
+  if (nrwlWorkspaceNode) {
+    packageJson.devDependencies['@nrwl/workspace'] =
+      nrwlWorkspaceNode.data.version;
+  }
+  writeJsonFile(`${options.outputPath}/package.json`, packageJson);
 }

--- a/packages/node/src/utils/generate-package-json.ts
+++ b/packages/node/src/utils/generate-package-json.ts
@@ -1,77 +1,17 @@
 import { ProjectGraph } from '@nrwl/workspace/src/core/project-graph';
-import {
-  readJsonFile,
-  writeJsonFile,
-} from '@nrwl/workspace/src/utilities/fileutils';
+import { writeJsonFile } from '@nrwl/workspace/src/utilities/fileutils';
 
 import { BuildNodeBuilderOptions } from './types';
+import { createPackageJson } from '@nrwl/workspace/src/utilities/create-package-json';
 import { OUT_FILENAME } from './config';
 
-/**
- * Creates a package.json in the output directory for support  to install dependencies within containers.
- *
- * If a package.json exists in the project, it will reuse that.
- *
- * @param projectName
- * @param graph
- * @param options
- * @constructor
- */
 export function generatePackageJson(
   projectName: string,
   graph: ProjectGraph,
   options: BuildNodeBuilderOptions
 ) {
-  const npmDeps = findAllNpmDeps(projectName, graph);
-  // default package.json if one does not exist
-  let packageJson = {
-    name: projectName,
-    version: '0.0.1',
-    main: OUT_FILENAME,
-    dependencies: {},
-  };
-
-  try {
-    packageJson = readJsonFile(`${options.projectRoot}/package.json`);
-    if (!packageJson.dependencies) {
-      packageJson.dependencies = {};
-    }
-  } catch (e) {}
-
-  const rootPackageJson = readJsonFile(`${options.root}/package.json`);
-
-  Object.entries(npmDeps).forEach(([packageName, version]) => {
-    // don't include devDeps
-    if (rootPackageJson.devDependencies?.[packageName]) {
-      return;
-    }
-
-    packageJson.dependencies[packageName] = version;
-  });
-
+  const packageJson = createPackageJson(projectName, graph, options);
+  packageJson.main = OUT_FILENAME;
+  delete packageJson.devDependencies;
   writeJsonFile(`${options.outputPath}/package.json`, packageJson);
-}
-
-function findAllNpmDeps(
-  projectName: string,
-  graph: ProjectGraph,
-  list: { [packageName: string]: string } = {},
-  seen = new Set<string>()
-) {
-  if (seen.has(projectName)) {
-    return list;
-  }
-
-  seen.add(projectName);
-
-  const node = graph.nodes[projectName];
-
-  if (node.type === 'npm') {
-    list[node.data.packageName] = node.data.version;
-  }
-  graph.dependencies[projectName]?.forEach((dep) => {
-    findAllNpmDeps(dep.target, graph, list, seen);
-  });
-
-  return list;
 }

--- a/packages/workspace/src/utilities/create-package-json.ts
+++ b/packages/workspace/src/utilities/create-package-json.ts
@@ -1,0 +1,106 @@
+import { ProjectGraph } from '../core/project-graph';
+import { readJsonFile } from './fileutils';
+
+/**
+ * Creates a package.json in the output directory for support to install dependencies within containers.
+ *
+ * If a package.json exists in the project, it will reuse that.
+ */
+export function createPackageJson(
+  projectName: string,
+  graph: ProjectGraph,
+  options: {
+    projectRoot?: string;
+    root?: string;
+  }
+): any {
+  const npmDeps = findAllNpmDeps(projectName, graph);
+  // default package.json if one does not exist
+  let packageJson = {
+    name: projectName,
+    version: '0.0.1',
+    dependencies: {},
+    devDependencies: {},
+  };
+  try {
+    packageJson = readJsonFile(`${options.projectRoot}/package.json`);
+    if (!packageJson.dependencies) {
+      packageJson.dependencies = {};
+    }
+    if (!packageJson.devDependencies) {
+      packageJson.devDependencies = {};
+    }
+  } catch (e) {}
+
+  const rootPackageJson = readJsonFile(`${options.root}/package.json`);
+  Object.entries(npmDeps).forEach(([packageName, version]) => {
+    if (rootPackageJson.devDependencies?.[packageName]) {
+      packageJson.devDependencies[packageName] = version;
+    } else {
+      packageJson.dependencies[packageName] = version;
+    }
+  });
+
+  return packageJson;
+}
+
+function findAllNpmDeps(
+  projectName: string,
+  graph: ProjectGraph,
+  list: { [packageName: string]: string } = {},
+  seen = new Set<string>()
+) {
+  if (seen.has(projectName)) {
+    return list;
+  }
+
+  seen.add(projectName);
+
+  const node = graph.nodes[projectName];
+
+  if (node.type === 'npm') {
+    list[node.data.packageName] = node.data.version;
+    recursivelyCollectPeerDependencies(node.name, graph, list);
+  }
+  graph.dependencies[projectName]?.forEach((dep) => {
+    findAllNpmDeps(dep.target, graph, list, seen);
+  });
+
+  return list;
+}
+
+function recursivelyCollectPeerDependencies(
+  projectName: string,
+  graph: ProjectGraph,
+  list: { [packageName: string]: string } = {},
+  seen = new Set<string>()
+) {
+  if (
+    !graph.nodes[projectName] ||
+    graph.nodes[projectName].type !== 'npm' ||
+    seen.has(projectName)
+  ) {
+    return list;
+  }
+
+  seen.add(projectName);
+  const packageName = graph.nodes[projectName].data.packageName;
+  try {
+    const packageJson = require(`${packageName}/package.json`);
+    if (!packageJson.peerDependencies) {
+      return list;
+    }
+
+    Object.keys(packageJson.peerDependencies)
+      .map((dependencyName) => `npm:${dependencyName}`)
+      .map((dependency) => graph.nodes[dependency])
+      .filter(Boolean)
+      .forEach((node) => {
+        list[node.data.packageName] = node.data.version;
+        recursivelyCollectPeerDependencies(node.name, graph, list, seen);
+      });
+    return list;
+  } catch (e) {
+    return list;
+  }
+}

